### ISSUE TITLE
Document `flutter run --machine` protocol

### ISSUE THREAD

**Issue by DanTup** [flutter/flutter#10546]
I'm building support for Flutter apps in Dart Code, but I'm struggling to find much info on the protocol for `flutter run --machine`. I found [this page](https://github.com/flutter/flutter/wiki/The-flutter-run-variants) but it just says "allows the use of JSON commands" without detailing what they are.

I presume it's a subset of [the daemon protocol](https://github.com/flutter/flutter/wiki/The-flutter-daemon-mode) but it would be good to know exactly which subset if so.

**Comment by eseidelGoogle** [comment]
This is a @devoncarew question. :)

**Comment by devoncarew** [comment]
It's largely the `app` domain from the protocol, but agree that we should improve the docs here.

**Comment by DanTup** [comment]
@devoncarew Does this mean things like `showMessage` won't come from `flutter run`? (not a bad thing; I have no good way to route them to the status bar / toasts!).

Some info on shutting down might be good too (I think when I just killed the process it didn't always close the app on the emulator, though shutting down gracefully is [causing me issues](https://github.com/Microsoft/vscode/issues/28192) too!).

**Comment by devoncarew** [comment]
> Does this mean things like showMessage won't come from flutter run? (not a bad thing; I have no good way to route them to the status bar / toasts!).

Yup, you won't see that from `flutter run --machine`.

> Some info on shutting down might be good too

Makes sense. For that, app.stop should work, and will terminate the process when finished. Killing the process should work as well - we catch the signal in flutter_tools, and kill the app on the device (the same as doing a ctrl-c on the cli).

> I think when I just killed the process it didn't always close the app on the emulator

If you are seeing this, we should open a bug - we do want the app to close on the device/emulator when the flutter_tools process goes away.

**Comment by DanTup** [comment]
I've raised #10556 about not always terminating cleanly.

Another quick question - if `wsUri` is not provided, what's the correct way to use `port`? Should I connect to `ws://127.0.0.1:${n.port}/ws`?

**Comment by devoncarew** [comment]
`wsUri` should always be provided - we used to only provide port, but switched to passing wsUri as well (which is the better field to pull from)

**Comment by DanTup** [comment]
Ok; the daemon page suggesting I might not get `wsUri` but if you changed long ago enough that we can assume everyone is newer, I'll just drop the `port` stuff.

**Comment by devoncarew** [comment]
Ah, yeah, I'll update that. There was a few week overlap w/ version skew for this field between clients and flutter_tools, but we're long past that.

**Comment by devoncarew** [comment]
(updated)

**Comment by DanTup** [comment]
Another thing worth documenting - which types of messages should be shown to the user? There's a lot of stuff coming back (`app.log`, `app.progress` and also non-JSON messages) but it's not clear which are things that are expected to be shown to the user and which are not (I don't want to just spam the output log junk).

I was perviously ignoring all of the non-JSON messages but then noticed that the "No devices connected." message comes through like that, so it I don't send them to the debug output, the user wouldn't know why the app is silently not launching (there appears to be no other output in this case).

**Comment by devoncarew** [comment]
makes sense re: updating the docs

For flutter run --machine, app.log messages should be shown as regular stdout, and app.progress messages should be translated into the platform specific progress UI. For vscode, this may mean having a flutter contribution to the status line (like we do for dart analysis)?

**Comment by DanTup** [comment]
@devoncarew I don't know how easily I can route back to the status bar from the debug adapter, but I'll have a dig.

You didn't mention the output that doesn't come wrapped in JSON - there are some strings just written raw to stdout like `No devices connected.` when trying to run with no devices - should these be filed as bugs and changed to json, or considered the same as `app.log`?

**Comment by devoncarew** [comment]
Things that aren't wrapped in json should just be passed through to the console's stdout.

Which items are you seeing as app.log events? Could you paste a few samples in here?

In terms of the progress display, I think we're just trying to get a good sense of progress to the users, esp. when things might take a while. If it's convenient to map that to an IDEs built-in progress mechanism, so much the better. Otherwise, perhaps just printing `<status message>...` at start, and `<status message> finished.` at stop?

**Comment by devoncarew** [comment]
(also, definitely very useful to have a 2nd implementor through this area; tells us a lot about what and how to document!)

**Comment by DanTup** [comment]
> Which items are you seeing as app.log events?

I dunno, you just said to output them in the console so I was agreeing. I was curious why there would be some sent as JSON (`app.log`) and some sent raw if they're expected to be handled the same way by integrators (I wondered if the raw output ones were just mistakes writing to output not considering `--machine`?).

> In terms of the progress display, I think we're just trying to get a good sense of progress to the users, esp. when things might take a while. If it's convenient to map that to an IDEs built-in progress mechanism, so much the better. 

Yeah, I'd like to put them in the status var if they've been crafted with that in mind. I'm trying to reduce the amount of stuff being spewed into the console output and I guess this stuff (which probably only matters for the short period of time while it's happening) is a good candidate for moving. I've opened https://github.com/Microsoft/vscode/issues/28551 to try and found out how best to get this to the status bar.

> (also, definitely very useful to have a 2nd implementor through this area; tells us a lot about what and how to document!)

:-) Is posting issues/ideas here (flutter GH issues) the preferred way to discuss these things?

**Comment by devoncarew** [comment]
Works for me. We can use the items here as notes the next time we update the protocol docs.

**Comment by DanTup** [comment]
It'd also be good to have some guidance on what needs to be done after a hot reload - it's not clear what state persists, needs resetting, etc.

For example, based on https://github.com/Dart-Code/Dart-Code/issues/318#issuecomment-310635478 it seems that breakpoints need to be re-sent. However it doesn't seem like Observatory is restarted; so are there other things I need to re-send to Observatory? (I'm actually wondering why I even need to re-send breakpoints - why are they being lost if Observatory is staying alive?)

**Comment by DanTup** [comment]
Based on info I can find (inc https://github.com/flutter/flutter-intellij/issues/273) it seems like only breakpoints will need resending. Can anyone confirm if that's right (eg I don't need to resend setLibraryDebuggable, exception pause modes, etc?)

**Comment by devoncarew** [comment]
Yeah, just breakpoints. The script that represents the old version of the library is still around in the VM, with the breakpoints associated with it that you had previously sent. You need to send new breakpoints for the reloaded library(ies), that will resolve against the new scripts for any reloaded libraries.

For IntelliJ, we choose to:
- pass in the pause after reload flag
- when we get the (newer) paused event (I believe PausePostRequest), we unset all the current breakpoints, reset all the breakpoints, and issue a resume for that isolate

**Comment by DanTup** [comment]
> The script that represents the old version of the library is still around in the VM, with the breakpoints associated with it that you had previously sent. 

Ah, this makes a little more sense now.

I did try hacking something together earlier but when I tried to send a `resume` I got back that the isolate wasn't paused. I wasn't handling the new event (I was assuming the response from the hot reload request was good enough) so maybe that's just a timing issue and switching to that event will resolve it.

**Comment by DanTup** [comment]
@devoncarew Is `PausePostRequest` only used for this? Is it valid for me to always clear/resend all breakpoints then a Resume when that happens (without having to track whether I'd just sent a restart request)? (I ask because the flutter code sending the request and the observatory code are somewhat disconnected)

**Comment by DanTup** [comment]
I added in stuff for re-sending breakpoints; but still hitting same problem. It seems like after the `PausePostRequest` the isolate is resuming itself (before I do it?!)

```text

[5:32:35 PM]: <== {"jsonrpc":"2.0","method":"streamNotify","params":{"streamId":"Debug","event":{"type":"Event","kind":"PausePostRequest","isolate":{"type":"@Isolate","fixedId":true,"id":"isolates\/467906369","name":"app.flx$main","number":"467906369"},"timestamp":1498235540430}}}
[5:32:35 PM]: ==> {"id":"474","method":"removeBreakpoint","params":{"isolateId":"isolates/467906369","breakpointId":"breakpoints/1"}}
[5:32:35 PM]: ==> {"id":"475","method":"addBreakpointWithScriptUri","params":{"isolateId":"isolates/467906369","scriptUri":"file:///m:/Coding/Applications/flutter/examples/stocks/lib/main.dart","line":65}}
[5:32:35 PM]: ==> {"id":"476","method":"removeBreakpoint","params":{"isolateId":"isolates/467906369","breakpointId":"breakpoints/2"}}
[5:32:35 PM]: ==> {"id":"477","method":"addBreakpointWithScriptUri","params":{"isolateId":"isolates/467906369","scriptUri":"package:stocks/main.dart","line":65}}

################ Why did this happen?
[5:32:35 PM]: <== {"jsonrpc":"2.0","method":"streamNotify","params":{"streamId":"Debug","event":{"type":"Event","kind":"Resume","isolate":{"type":"@Isolate","fixedId":true,"id":"isolates\/467906369","name":"app.flx$main","number":"467906369"},"timestamp":1498235536850}}}

[5:32:35 PM]: <== {"jsonrpc":"2.0","method":"streamNotify","params":{"streamId":"Debug","event":{"type":"Event","kind":"BreakpointRemoved","isolate":{"type":"@Isolate","fixedId":true,"id":"isolates\/467906369","name":"app.flx$main","number":"467906369"},"timestamp":1498235540433,"breakpoint":{"type":"Breakpoint","fixedId":true,"id":"breakpoints\/1","breakpointNumber":1,"resolved":true,"location":{"type":"SourceLocation","script":{"type":"@Script","id":"objects\/154","uri":"file:\/\/\/m:\/Coding\/Applications\/flutter\/examples\/stocks\/lib\/main.dart","_kind":"script"},"tokenPos":350}}}}}
[5:32:35 PM]: <== {"jsonrpc":"2.0", "result":{"type":"Success"},"id":"474"}
[5:32:35 PM]: <== {"jsonrpc":"2.0","method":"streamNotify","params":{"streamId":"Debug","event":{"type":"Event","kind":"BreakpointAdded","isolate":{"type":"@Isolate","fixedId":true,"id":"isolates\/467906369","name":"app.flx$main","number":"467906369"},"timestamp":1498235540440,"breakpoint":{"type":"Breakpoint","fixedId":true,"id":"breakpoints\/3","breakpointNumber":3,"resolved":false,"location":{"type":"UnresolvedSourceLocation","scriptUri":"file:\/\/\/m:\/Coding\/Applications\/flutter\/examples\/stocks\/lib\/main.dart","line":65}}}}}
[5:32:35 PM]: <== {"jsonrpc":"2.0", "result":{"type":"Breakpoint","fixedId":true,"id":"breakpoints\/3","breakpointNumber":3,"resolved":false,"location":{"type":"UnresolvedSourceLocation","scriptUri":"file:\/\/\/m:\/Coding\/Applications\/flutter\/examples\/stocks\/lib\/main.dart","line":65}},"id":"475"}
[5:32:35 PM]: <== {"jsonrpc":"2.0","method":"streamNotify","params":{"streamId":"Debug","event":{"type":"Event","kind":"BreakpointRemoved","isolate":{"type":"@Isolate","fixedId":true,"id":"isolates\/467906369","name":"app.flx$main","number":"467906369"},"timestamp":1498235540441,"breakpoint":{"type":"Breakpoint","fixedId":true,"id":"breakpoints\/2","breakpointNumber":2,"resolved":false,"location":{"type":"UnresolvedSourceLocation","scriptUri":"package:stocks\/main.dart","line":65}}}}}
[5:32:35 PM]: <== {"jsonrpc":"2.0", "result":{"type":"Success"},"id":"476"}
[5:32:35 PM]: <== {"jsonrpc":"2.0","method":"streamNotify","params":{"streamId":"Debug","event":{"type":"Event","kind":"BreakpointAdded","isolate":{"type":"@Isolate","fixedId":true,"id":"isolates\/467906369","name":"app.flx$main","number":"467906369"},"timestamp":1498235540445,"breakpoint":{"type":"Breakpoint","fixedId":true,"id":"breakpoints\/4","breakpointNumber":4,"resolved":false,"location":{"type":"UnresolvedSourceLocation","scriptUri":"package:stocks\/main.dart","line":65}}}}}
[5:32:35 PM]: <== {"jsonrpc":"2.0", "result":{"type":"Breakpoint","fixedId":true,"id":"breakpoints\/4","breakpointNumber":4,"resolved":false,"location":{"type":"UnresolvedSourceLocation","scriptUri":"package:stocks\/main.dart","line":65}},"id":"477"}
[5:32:35 PM]: ==> {"id":"478","method":"resume","params":{"isolateId":"isolates/467906369"}}

```

**Comment by devoncarew** [comment]
The `PausePostRequest` is only sent when an isolate pauses as a result of a reload (w/ the `pause`: true param). It _is_ a type of a pause, so I'd very very surprised if the isolate unpauses itself automatically :) Not clear why it's resuming above. There are no other clients connected to the VM?

**Comment by DanTup** [comment]
I don't believe so; but I'll have a dig through to ensure there's no way I could be opening a second connection. I've tried restarting everything and pulling latest Flutter code but that didn't help.

**Comment by DanTup** [comment]
I added logging to the code that connects to observatory and it does seem to only be being done once :(

**Comment by DanTup** [comment]
@devoncarew I can repro the same behaviour in a simple script.. I'll tidy it up and raise an issue.

**Comment by DanTup** [comment]
I've raised #10934 with a script and logged output showing this behaviour (the script is really simple; it just waits for 10 secs after the app restarts and issues a reload, which results in a `PausePostRequest` immediately followed by a `Resume`.

**Comment by DanTup** [comment]
@devoncarew Just implementing a command to open the timeline (https://github.com/Dart-Code/Dart-Code/issues/381).. Looks like currently we get `{ip}:{port}/ws` back as the `wsUri` from `app.debugPort` but to open Observatory/Timeline I need to take that off.

Is it safe for me to just check if it ends with `/ws` and remove that (`uri.substring(0, uri.length - 3)`) or is there the possibility this format will change in the future? (for ex., would I be better stripping back to the root?)

**Comment by DanTup** [comment]
Oh, and I need to change the protocol from `ws` to `http` too. Would it better if the browser-friendly Observatory address was provided to avoid clients having to do parsing?

(For now I'm just implementing it as described above, seems like the chances of it changing are slim).

**Comment by devoncarew** [comment]
You should also get back a `baseUri` field as well. You can use that, and won't have to rely as much in munging the `ws:` url.

**Comment by DanTup** [comment]
@devoncarew Isn't `baseUri` different; used for breakpoints on the device?

> "baseUri":"file:///Users/danny/Library/Developer/CoreSimulator/Devices/C76BB178-2046-4302-B25B-0A747ECAA276/data/Containers/Data/Application/CCFE1472-7C21-4A5B-B733-022D49ADEEEB/tmp/test_breakpointskIQNRi/test_breakpoints/"

**Comment by devoncarew** [comment]
Ah, yeah, absolutely. It's been a while since I was in the code; `baseUri` should be used for breakpoints, and your best bet for building an observatory url is using `wsUri`: https://github.com/flutter/flutter-intellij/blob/ebf310549c5282a3db257cac2c16608b9fe7233b/src/io/flutter/run/daemon/FlutterApp.java#L98

**Comment by DanTup** [comment]
@devoncarew Cool, thanks for confirming. Seems like my implementation is pretty much the same as IntelliJ, so should be ok :-)

```ts
if (uri.endsWith('/ws')) {
	var browserFriendlyUri = uri.substring(0, uri.length - 3);
	if (browserFriendlyUri.startsWith('ws:'))
		browserFriendlyUri = "http:" + browserFriendlyUri.substring(3);
	this.sendEvent(new Event("dart.observatoryUri", { observatoryUri: browserFriendlyUri.toString() }));
}
```

**Comment by DanTup** [comment]
I've also updated the note on the [`flutter run`](https://github.com/flutter/flutter/wiki/The-flutter-run-variants) to have a link to the daemon page.
